CMakeLists.txt

```diff
@@ -250,41 +193,79 @@ add_executable(
     NEMS/src/nems_c_rusage.c
     NEMS/src/ENS_Cpl/ENS_CplComp_ESMFMod_STUB.F90
 )
+add_dependencies(ufs-weather-model-lib fv3atm)
+target_link_libraries(ufs-weather-model-lib PRIVATE esmf)
 
-target_compile_definitions(NEMS.exe PRIVATE -DESMF_VERSION_MAJOR=${ESMF_VERSION_MAJOR})
-target_compile_definitions(NEMS.exe PRIVATE -DFRONT_FMS)
-target_compile_definitions(NEMS.exe PRIVATE -DFRONT_FV3=fv3gfs_cap_mod)
+target_compile_definitions(ufs-weather-model-lib PRIVATE -DESMF_VERSION_MAJOR=${ESMF_VERSION_MAJOR})
+target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_FMS)
+target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_FV3=fv3gfs_cap_mod)
+
+target_include_directories(ufs-weather-model-lib PRIVATE ${PROJECT_BINARY_DIR}/NEMS/src)
+target_include_directories(ufs-weather-model-lib PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/FV3/mod)
+target_include_directories(ufs-weather-model-lib PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/FMS/mod)
+
+if(WW3)
+    target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_WW3=WMESMFMD)
+    target_include_directories(ufs-weather-model-lib PRIVATE ${PROJECT_SOURCE_DIR}/WW3/model/mod)
+    set_target_properties(ufs-weather-model-lib PROPERTIES Fortran_MODULE_DIRECTORY ${PROJECT_SOURCE_DIR}/WW3/model/mod)
+    add_dependencies(ufs-weather-model-lib ww3_nems)
+endif()
 
-target_include_directories(NEMS.exe PRIVATE ${ESMF_MOD} )
+set_target_properties(ufs-weather-model-lib PROPERTIES Fortran_MODULE_DIRECTORY ${PROJECT_BINARY_DIR}/mod)
+target_include_directories(ufs-weather-model-lib PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/mod>
+                                       $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/FV3/mod>
+                                       $<INSTALL_INTERFACE:${CMAKE_INSTALL_PREFIX}/mod>)
+
+configure_file(${PROJECT_SOURCE_DIR}/NEMS/src/ESMFVersionDefine_ESMF_NUOPC.h
+               ${PROJECT_BINARY_DIR}/NEMS/src/ESMFVersionDefine.h
+               COPYONLY)
+# until we remove "../ESMFVersionDefine.h" from ENS_Cpl/ENS_CplComp_ESMFMod_STUB.F90
+configure_file(${PROJECT_SOURCE_DIR}/NEMS/src/ESMFVersionDefine_ESMF_NUOPC.h
+               ${PROJECT_BINARY_DIR}/NEMS/ESMFVersionDefine.h
+               COPYONLY)
+configure_file(${PROJECT_SOURCE_DIR}/NEMS/src/ESMFConvenienceMacros.h
+               ${PROJECT_BINARY_DIR}/NEMS/src/ESMFConvenienceMacros.h
+               COPYONLY)
+
+target_link_libraries(ufs-weather-model-lib PUBLIC fv3atm)
+
+###############################################################################
+### UFS executable
+###############################################################################
+add_executable(NEMS.exe
+               NEMS/src/MAIN_NEMS.F90
+)
+add_dependencies(NEMS.exe ufs-weather-model-lib)
+
+target_compile_definitions(NEMS.exe PRIVATE -DESMF_VERSION_MAJOR=${ESMF_VERSION_MAJOR})
```

Comment:
Is this needed? Wouldn't it inherit from `ufs-weather-model-lib`?

Reply:
It is needed at this moment, because `ESMF_VERSION_MAJOR` is added as `PRIVATE` definition to `ufs-weather-model-lib`. We can add it as `PUBLIC` but then all users of ufs-weather-model-lib (ie. JEDI) will get it. Do we want that? I thought it makes more sense to keep it private. In any case this will all be replaced pretty soon.